### PR TITLE
Add binary-compatibility-validator and API dumps for desktop

### PR DESCRIPTION
**Pull request by igordmn**
https://github.com/Kotlin/binary-compatibility-validator

This plugin dumps all API for JVM modules to files (in the future - for other targets). If API changed, we should call:
```
./gradlew desktopApiDump
```
to commit a new API, otherwise CI will fail (`./gradlew desktopCICheck` will be called on CI`).

If you see that some strings are deleted in the API dump, most probably it means an incompatible API change (the plugin doesn't check itself, unfortunately).

See [the official workflow guide](https://github.com/Kotlin/binary-compatibility-validator#workflow) to understand it more.

Example of error:
```
Execution failed for task ':compose:material3:material3:desktopApiCheck'.
> API check failed for project material3.
  --- D:\Work\compose-multiplatform-core\compose\material3\material3\api\desktop\material3.api
  +++ D:\Work\compose-multiplatform-core\out\androidx\compose\material3\material3\build\api\desktop\material3.api
  @@ -552,6 +552,11 @@
   public abstract interface annotation class androidx/compose/material3/ExperimentalMaterial3Api : java/lang/annotation/Annotation {
   }

  +public final class androidx/compose/material3/FF {
  +     public static final field $stable I
  +     public fun <init> ()V
  +}
  +
   public final class androidx/compose/material3/FabPosition {
        public static final field Companion Landroidx/compose/material3/FabPosition$Companion;
        public static final synthetic fun box-impl (I)Landroidx/compose/material3/FabPosition;

   You can run :material3:apiDump task to overwrite API declarations
```

One disadvantage that error shows `apiDump` instead of `desktopApiDump`. `apiDump` doesn't work because of the broken Android target, and to fix it properly we need to do a lot of changes. Just disabling the tasks doesn't work.

- Desktop API is constructed from desktopMain + skikoMain + commonMain sourceSets

- Android API is constructed from desktopMain + commonMain. It is disabled for now, because Android build is broken. The plan is to fix the Android target, and dump Android API as well. It will help us to track new API that we need to port more easily.

P.S. Jetpack Compose sources has its own API generator (Metalava), but it won't be applicable to other Kotlin targets in the future, but binary-compatibility-validator [will](https://github.com/Kotlin/binary-compatibility-validator/issues/149)

## Test
1. call desktopCICheck - it succeeds
2. change/add API
3. call desktopCICheck - it fails

### CODE REVIEW

File: compose/material3/material3/api/desktop/material3.api [hunk 1]
```diff
@@ -0,0 +1,1370 @@
...
+	public final fun getIconContentColor (Landroidx/compose/runtime/Composer;I)J
+	public final fun getShape (Landroidx/compose/runtime/Composer;I)Landroidx/compose/ui/graphics/Shape;
+	public final fun getTextContentColor (Landroidx/compose/runtime/Composer;I)J
+	public final fun getTitleContentColor (Landroidx/compose/runtime/Composer;I)J
+	public final fun getTonalElevation-D9Ej5fM ()F
+}
+
+public final class androidx/compose/material3/AlertDialog_skikoKt {
+	public static final fun AlertDialog (Lkotlin/jvm/functions/Function0;Landroidx/compose/ui/Modifier;Landroidx/compose/ui/window/DialogProperties;Lkotlin/jvm/functions/Function2;Landroidx/compose/runtime/Composer;II)V
+	public static final fun AlertDialog-Oix01E0 (Lkotlin/jvm/functions/Function0;Lkotlin/jvm/functions/Function2;Landroidx/compose/ui/Modifier;Lkotlin/jvm/functions/Function2;Lkotlin/jvm/functions/Function2;Lkotlin/jvm/functions/Function2;Lkotlin/jvm/functions/Function2;Landroidx/compose/ui/graphics/Shape;JJJJFLandroidx/compose/ui/window/DialogProperties;Landroidx/compose/runtime/Composer;III)V
```
Comment (MatkovIvan) [inline compose/material3/material3/api/desktop/material3.api:14]: 1. What `-Oix01E0` mean here?
2. It looks like this validator doesn't handle values of default parameters, right?

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -161,6 +162,7 @@ gradleIncapHelper = { module = "net.ltgt.gradle.incap:incap", version.ref = "inc
 gradleIncapHelperProcessor = { module = "net.ltgt.gradle.incap:incap-processor", version.ref = "incap" }
 kotlinAnnotationProcessingEmbeddable = { module = "org.jetbrains.kotlin:kotlin-annotation-processing-embeddable", version.ref = "kotlin" }
 kotlinBenchmarkRuntime = { module = "org.jetbrains.kotlinx:kotlinx-benchmark-runtime", version.ref = "kotlinBenchmark" }
+kotlinBinaryCompatibilityValidatorGradlePluginz = { module = "org.jetbrains.kotlinx.binary-compatibility-validator:org.jetbrains.kotlinx.binary-compatibility-validator.gradle.plugin", version.ref = "kotlinBinaryCompatibilityValidator" }
```
Comment (MatkovIvan) [inline gradle/libs.versions.toml]:
```suggestion
kotlinBinaryCompatibilityValidatorGradlePlugin = { module = "org.jetbrains.kotlinx.binary-compatibility-validator:org.jetbrains.kotlinx.binary-compatibility-validator.gradle.plugin", version.ref = "kotlinBinaryCompatibilityValidator" }
```

File: buildSrc/shared-dependencies.gradle [hunk 3]
```diff
@@ -18,6 +18,7 @@ dependencies {
     implementation(libs.androidToolsRepository) // com.android.repository for Stable AIDL plugin
     implementation(libs.androidToolsSdkCommon) // com.android.ide.common for Stable AIDL plugin
     implementation(libs.kotlinGradlePluginz)
+    implementation(libs.kotlinBinaryCompatibilityValidatorGradlePluginz)
```
Comment (MatkovIvan) [inline buildSrc/shared-dependencies.gradle]:
```suggestion
    implementation(libs.kotlinBinaryCompatibilityValidatorGradlePlugin)
```

Comment (igordmn) [reply buildSrc/shared-dependencies.gradle]: This was a confusing name, when I was writing it :) (I kept the pattern of other plugins). But let's change it.

File: compose/material3/material3/api/desktop/material3.api (hunk 1 above)
Comment (igordmn) [reply compose/material3/material3/api/desktop/material3.api:14]: 1. It is how Kotlin compiles a fun with value classes (Color, for example):
image

2. Kotlin compiles such functions to `Foo$default`:

image

Comment (igordmn) [reply compose/material3/material3/api/desktop/material3.api:14]: 2. And it seems, yes, we can't track default values changes. But it is a behaviour change, not API change

**Review by MatkovIvan — APPROVED** [review]

**Comment by igordmn** [comment]
Added a readme

**Review by Walingar — APPROVED** [review]

File: MULTIPLATFORM.md [hunk 4]
```diff
@@ -26,6 +26,38 @@ Run tests for UIKit:
 ./gradlew :mpp:testUIKit
 ```
 
+### API checks
+Compose Multiplatform stores all public API in *.api files. If any API is added/changed, `./gradlew desktopCICheck` will fail with an error that API is changed (it runs on CI). Example:
```
Comment (Walingar) [inline MULTIPLATFORM.md]: desktopCICheck -> checkDesktop

Comment (igordmn) [reply MULTIPLATFORM.md]: Fixed

**Pull request merged**